bindings/pubsubadapter/pubsubadapter.go

```diff
@@ -0,0 +1,150 @@
+// the topic is not overridden in the request metadata.
+func WithInvokeTopicAttribute(invokeTopicAttribute string) Option {
+	return func(a *Adapter) {
+		a.invokeTopicAttribute = invokeTopicAttribute
+	}
+}
+
+// WithInvokeTopicMetadata overrides the default invoke topic metadata
+// name that determines the metadata property that overrides the topic
+// to publish to when Invoke is called.
+func WithInvokeTopicMetadata(invokeTopicMetadata string) Option {
+	return func(a *Adapter) {
+		a.invokeTopicMetadata = invokeTopicMetadata
+	}
+}
+
+// New returns an input/output binding adapter for `broker`.
+func New(broker pubsub.PubSub, options ...Option) *Adapter {
+	a := Adapter{ //nolint:exhaustivestruct
+		broker:               broker,
+		readTopicAttribute:   defaultReadTopicAttribute,
+		invokeTopicAttribute: defaultInvokeTopicAttribute,
+		invokeTopicMetadata:  defaultInvokeTopicMetadata,
+	}
+
+	for _, opt := range options {
+		opt(&a)
+	}
+
+	return &a
+}
+
+// Init reads the topic property and passes the properties
+// to the underlying PubSub component.
+func (a *Adapter) Init(metadata bindings.Metadata) error {
+	if t, ok := metadata.Properties[a.readTopicAttribute]; ok && t != "" {
+		a.readTopic = t
+	} else {
+		return fmt.Errorf("pubsub adapter: %q attribute was missing", a.readTopicAttribute)
+	}
+
+	if t, ok := metadata.Properties[a.invokeTopicAttribute]; ok && t != "" {
+		a.invokeTopic = t
+	} else {
+		return fmt.Errorf("pubsub adapter: %q attribute was missing", a.invokeTopicAttribute)
+	}
+
+	a.name = metadata.Name
+
+	return a.broker.Init(pubsub.Metadata{ //nolint:wrapcheck
+		Properties: metadata.Properties,
+	})
+}
+
+// Read subscribes to the configured topic and invokes `handler`
+// when a message is received.
+func (a *Adapter) Read(handler func(*bindings.ReadResponse) ([]byte, error)) error {
+	return a.broker.Subscribe(pubsub.SubscribeRequest{ //nolint:wrapcheck
+		Topic:    a.invokeTopic,
+		Metadata: map[string]string{},
+	}, func(ctx context.Context, msg *pubsub.NewMessage) error {
+		_, err := handler(&bindings.ReadResponse{
+			Data:     msg.Data,
+			Metadata: msg.Metadata,
+		})
+
+		return err
+	})
+}
+
+// Invoke publishes a message to the configured topic or
+// optionally the topic indicated in `req.Metadata`.
+func (a *Adapter) Invoke(req *bindings.InvokeRequest) (*bindings.InvokeResponse, error) {
+	topic := a.invokeTopic
+	if t, ok := req.Metadata[a.invokeTopicMetadata]; ok && t != "" {
+		topic = t
+	}
+
+	err := a.broker.Publish(&pubsub.PublishRequest{
+		Data:       req.Data,
+		PubsubName: a.name,
+		Topic:      topic,
+		Metadata:   req.Metadata,
+	})
+
+	return nil, err //nolint:wrapcheck
+}
+
+// Operations returns "create" and "publish" as an operations
+// but operation is ignored by `Invoke`.
+func (a *Adapter) Operations() []bindings.OperationKind {
+	return []bindings.OperationKind{bindings.CreateOperation, bindings.OperationKind("publish")}
```

Comment:
This suggests a broader problem that should be standardized across bindings: implementations should generally provide a broader range of consistent actions (i.e. other pubsub-like bindings should also advertise `publish`), but I suspect that most of them use `create` as publish to fit with how the conformance test behaves, which is an undesirable restriction.

Reply:
Agreed.  For now, I'm trying not to break the existing bindings, however since non are "stable", we could make Invoke inspect the value of the operation.